File: cpp/src/ensoul.cpp
#include "ensoul/ensoul.hpp"

namespace ensoul {

EnsoulClient::EnsoulClient(
    const std::string& api_key,
    const std::string& base_url,
    const std::string& bearer_token,
    long timeout_ms,
    int max_retries,
    const std::map<std::string, std::string>& custom_headers)
{
    std::string resolved_key = api_key;
    if (resolved_key.empty()) {
        const char* env_key = std::getenv("ENSOUL_API_KEY");
        if (env_key) resolved_key = env_key;
    }

    std::string resolved_url = base_url;
    if (resolved_url == DEFAULT_BASE_URL) {
        const char* env_url = std::getenv("ENSOUL_BASE_URL");
        if (env_url) resolved_url = env_url;
    }

    config_.base_url = resolved_url;
    config_.api_key = resolved_key;
    config_.bearer_token = bearer_token;
    config_.timeout_ms = timeout_ms;
    config_.max_retries = max_retries;
    config_.custom_headers = custom_headers;

    transport_ = std::make_unique<HttpClient>(config_);
    init_resources();
}

EnsoulClient::EnsoulClient(const ClientConfig& config,
                           std::unique_ptr<IHttpTransport> transport)
    : config_(config), transport_(std::move(transport))
{
    init_resources();
}

EnsoulClient EnsoulClient::with_http_client(
    const ClientConfig& config,
    std::unique_ptr<IHttpTransport> transport)
{
    return EnsoulClient(config, std::move(transport));
}

void EnsoulClient::init_resources() {
    personas_ = std::make_unique<Personas>(*transport_);
    chat_ = std::make_unique<ChatResource>(*transport_);
    domains_ = std::make_unique<DomainsResource>(*transport_);
    simulations_ = std::make_unique<SimulationsResource>(*transport_);
    aggregate_ = std::make_unique<AggregateResource>(*transport_);
    memory_ = std::make_unique<MemoryResource>(*transport_);
    sessions_ = std::make_unique<SessionsResource>(*transport_);
    frameworks_ = std::make_unique<FrameworksResource>(*transport_);
    auth_ = std::make_unique<AuthResourceNS>(*transport_);
    health_ = std::make_unique<HealthResource>(*transport_);
    info_ = std::make_unique<InfoResource>(*transport_);
    audit_ = std::make_unique<AuditResource>(*transport_);
}

} // namespace ensoul

File: swift/Sources/Ensoul/Ensoul.swift
/// Ensoul SDK for Swift.
///
/// Usage:
/// ```swift
/// let client = EnsoulClient(apiKey: "sk_...")
/// let persona = try await client.personas.create(name: "Test", domain: "my_domain")
/// let response = try await client.chat.send(personaId: persona.id, message: "Hello!")
/// ```
///
/// The API key can also be set via the `ENSOUL_API_KEY` environment variable.
import Foundation

@available(iOS 15.0, macOS 12.0, *)
public class EnsoulClient {
    public static let version = "0.1.0"

    // MARK: - Resource namespaces

    public let personas: Personas
    public let chat: Chat
    public let domains: Domains
    public let simulations: Simulations
    public let aggregate: Aggregate
    public let memory: Memory
    public let sessions: Sessions
    public let frameworks: Frameworks
    public let auth: AuthResource
    public let health: Health
    public let info: Info
    public let audit: Audit

    private let httpClient: HTTPClient

    // MARK: - Public init

    /// Create an Ensoul client.
    ///
    /// - Parameters:
    ///   - apiKey: API key for authentication. Falls back to `ENSOUL_API_KEY` env var.
    ///   - baseURL: API base URL. Defaults to `"https://api.ensoul-ai.com"`.
    ///   - bearerToken: OAuth2 bearer token (alternative to API key).
    ///   - timeout: Request timeout in seconds. Defaults to 30.
    ///   - maxRetries: Maximum retry attempts. Defaults to 2.
    ///   - customHeaders: Additional headers merged into every request.
    ///   - session: Custom `URLSession` (useful for injecting `MockURLProtocol` in tests).
    public init(
        apiKey: String? = nil,
        baseURL: String? = nil,
        bearerToken: String? = nil,
        timeout: TimeInterval = defaultTimeout,
        maxRetries: Int = 2,
        customHeaders: [String: String] = [:],
        session: URLSession = .shared
    ) {
        let resolvedAPIKey = apiKey ?? ProcessInfo.processInfo.environment["ENSOUL_API_KEY"]
        let resolvedBaseURL = baseURL ?? ProcessInfo.processInfo.environment["ENSOUL_BASE_URL"] ?? "https://api.ensoul-ai.com"

        let config = ClientConfig(
            baseURL: resolvedBaseURL,
            apiKey: resolvedAPIKey,
            bearerToken: bearerToken,
            timeout: timeout,
            maxRetries: maxRetries,
            customHeaders: customHeaders
        )

        self.httpClient = HTTPClient(config: config, session: session)

        self.personas    = Personas(client: httpClient)
        self.chat        = Chat(client: httpClient)
        self.domains     = Domains(client: httpClient)
        self.simulations = Simulations(client: httpClient)
        self.aggregate   = Aggregate(client: httpClient)
        self.memory      = Memory(client: httpClient)
        self.sessions    = Sessions(client: httpClient)
        self.frameworks  = Frameworks(client: httpClient)
        self.auth        = AuthResource(client: httpClient)
        self.health      = Health(client: httpClient)
        self.info        = Info(client: httpClient)
        self.audit       = Audit(client: httpClient)
    }

    // MARK: - Internal init (for testing with a pre-built HTTPClient)

    /// Create a client with a custom `HTTPClient` (for unit tests that inject a mock transport).
    internal init(httpClient: HTTPClient) {
        self.httpClient  = httpClient
        self.personas    = Personas(client: httpClient)
        self.chat        = Chat(client: httpClient)
        self.domains     = Domains(client: httpClient)
        self.simulations = Simulations(client: httpClient)
        self.aggregate   = Aggregate(client: httpClient)
        self.memory      = Memory(client: httpClient)
        self.sessions    = Sessions(client: httpClient)
        self.frameworks  = Frameworks(client: httpClient)
        self.auth        = AuthResource(client: httpClient)
        self.health      = Health(client: httpClient)
        self.info        = Info(client: httpClient)
        self.audit       = Audit(client: httpClient)
    }
}

// MARK: - HTTPClient body-label convenience overloads
//
// Resource files (Personas.swift, Chat.swift, Simulations.swift) call HTTPClient
// methods with a `body:` parameter label; the core HTTPClient uses `json:`.
// These extensions bridge the gap so the module compiles.

@available(iOS 15.0, macOS 12.0, *)
extension HTTPClient {
    func post(_ path: String, body: [String: Any]? = nil) async throws -> (Data, HTTPURLResponse) {
        return try await post(path, json: body)
    }

    func put(_ path: String, body: [String: Any]? = nil) async throws -> (Data, HTTPURLResponse) {
        return try await put(path, json: body)
    }

    func patch(_ path: String, body: [String: Any]? = nil) async throws -> (Data, HTTPURLResponse) {
        return try await patch(path, json: body)
    }

    func streamSSE(method: String, path: String, body: [String: Any]?) -> SSEStream {
        return streamSSE(method: method, path: path, json: body)
    }
}

